plugins/image-prioritizer/class-image-prioritizer-background-image-styled-tag-visitor.php

```diff
@@ -65,33 +79,126 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 
 		// If this element is the LCP (for a breakpoint group), add a preload link for it.
 		foreach ( $context->url_metric_group_collection->get_groups_by_lcp_element( $xpath ) as $group ) {
-			$link_attributes = array(
-				'rel'           => 'preload',
-				'fetchpriority' => 'high',
-				'as'            => 'image',
-				'href'          => $background_image_url,
-				'media'         => 'screen',
-			);
-
-			$context->link_collection->add_link(
-				$link_attributes,
-				$group->get_minimum_viewport_width(),
-				$group->get_maximum_viewport_width()
-			);
+			$this->add_preload_link( $context->link_collection, $group, $background_image_url );
 		}
 
 		$this->lazy_load_bg_images( $context );
 
 		return true;
 	}
 
+	/**
+	 * Gets the common LCP element external background image for a URL Metric group.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_URL_Metric_Group $group Group.
+	 * @return LcpElementExternalBackgroundImage|null
+	 */
+	private function get_common_lcp_element_external_background_image( OD_URL_Metric_Group $group ): ?array {
+
+		// If the group is not fully populated, we don't have enough URL Metrics to reliably know whether the background image is consistent across page loads.
+		// This is intentionally not using $group->is_complete() because we still will use stale URL Metrics in the calculation.
+		if ( $group->count() !== $group->get_sample_size() ) {
+			return null;
+		}
+
+		$previous_lcp_element_external_background_image = null;
+		foreach ( $group as $url_metric ) {
+			/**
+			 * Stored data.
+			 *
+			 * @var LcpElementExternalBackgroundImage|null $lcp_element_external_background_image
+			 */
+			$lcp_element_external_background_image = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+			if ( ! is_array( $lcp_element_external_background_image ) ) {
+				return null;
+			}
+			if ( null !== $previous_lcp_element_external_background_image && $previous_lcp_element_external_background_image !== $lcp_element_external_background_image ) {
+				return null;
+			}
+			$previous_lcp_element_external_background_image = $lcp_element_external_background_image;
+		}
+
+		return $previous_lcp_element_external_background_image;
+	}
+
+	/**
+	 * Maybe preloads external background image.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Context.
+	 */
+	private function maybe_preload_external_lcp_background_image( OD_Tag_Visitor_Context $context ): void {
+		// Gather the tuples of URL Metric group and the common LCP element external background image.
+		if ( ! is_array( $this->group_common_lcp_element_external_background_images ) ) {
+			$this->group_common_lcp_element_external_background_images = array();
+			foreach ( $context->url_metric_group_collection as $group ) {
+				$common = $this->get_common_lcp_element_external_background_image( $group );
+				if ( is_array( $common ) ) {
+					$this->group_common_lcp_element_external_background_images[] = array( $group, $common );
+				}
+			}
+		}
```

Comment:
Why is this being written to a class property, rather than just a local variable? Is it for memoization? If so, why would this method be called multiple times for the same context?

Reply:
Yes, it is for memoization. This class is a tag visitor which is invoked for every tag in the document. So above in the `__invoke()` method there is this logic:

```php
		if ( is_null( $background_image_url ) ) {
			$this->maybe_preload_external_lcp_background_image( $context );
			return false;
		}
```

So the first time that the tag visitor is invoked, it populates this class variable with the data from the collection. If there is are LCP background image elements for the viewport groups, then immediately below it short-circuits:

```
		// There are no common LCP background images, so abort.
		if ( count( $this->group_common_lcp_element_external_background_images ) === 0 ) {
			return;
		}
```

Otherwise it then proceeds to check if the current tag we're visiting is the same tag that was reported as the LCP element which had a background image, and if so, it adds the preload link (further below).

Reply:
Makes sense.

What throws me off a bit is that the memoization depends on a value that comes from the `$context` but it is memoized independently of that `$context`. I assume the `$group`s would always be the same across all contexts for a page? If so, this makes sense as is. My hesitation comes from the technical dependency on `$context` but not "key-ing" the memoized value with it. So I think we should at least add a comment on why this is not needed here.

Reply:
Correct, the groups of URL Metrics do not change while visiting the tags in the document.

How about this: https://github.com/WordPress/performance/pull/1697/files#r1866624177